lib/stages/VisSharedMemWriter.cpp

```diff
@@ -0,0 +1,379 @@
+    uint64_t fpga_seq = frame.metadata()->fpga_seq_start;
+
+
+    if (wait_for_semaphore() == false) {
+        ERROR("Failed to acquire semaphore {} before metadata was updated.\n", _sem_name);
+        return;
+    }
+
+    DEBUG("Writing fpga_seq {} to time index {}\n", fpga_seq, time_ind);
+    memcpy(access_record_write_pos, &fpga_seq, sizeof(fpga_seq));
+
+    // update num_writes
+    num_writes++;
+    memcpy(structured_data_addr, &num_writes, sizeof(num_writes));
+
+    if (sem_post(sem) == -1) {
+        FATAL_ERROR("Failed to release semaphore {}", _sem_name);
+        return;
+    }
+
+}
+
+void VisSharedMemWriter::main_thread() {
+    DEBUG("Reached main thread\n");
+
+    frameID frame_id(in_buf);
+
+    // The current position in the ring buffer of the most recent time sample
+    // from 0 -> _ntime
+    cur_pos = modulo<int>(_ntime);
+
+    // Create the semaphore, and gain first access to it
+    sem = sem_open(
+            _sem_name.c_str(),
+            (O_CREAT | O_EXCL),
+            (S_IRUSR | S_IWUSR),
+            1
+        );
+
+    if (sem == SEM_FAILED) {
+        FATAL_ERROR("Failed to create semaphore {}", _sem_name);
+        return;
+    }
+
+    DEBUG("Semaphore created.\n");
+
+    // Acquire semaphore until shared memory is created
+    if (wait_for_semaphore() == false) {
+        ERROR("Failed to acquire semaphore {}", _sem_name);
+        return;
+    }
+
+    // Set up the structure of the ring buffer shared memory
+    // Get one frame for reference
+    wait_for_full_frame(in_buf, unique_name.c_str(), frame_id);
+
+    auto frame = visFrameView(in_buf, frame_id);
+
+    // Build the frequency index
+    std::map<uint32_t, uint32_t> freq_id_map;
+    auto& dm = datasetManager::instance();
+
+    const freqState* freq_state = dm.dataset_state<freqState>(frame.dataset_id);
+
+    uint ind = 0;
+    for (auto& f : freq_state->get_freqs())
+        freq_id_map[f.first] = ind++;
+
+
+    // Figure out the ring buffer structure
+    nfreq = freq_state->get_freqs().size();
+
+    // Set the alignment (in kB)
+    size_t alignment = 4096; // Align on page boundaries
+
+    // Calculate the ring buffer structure
+
+    data_size = frame.data_size;
+    metadata_size = sizeof(visMetadata);
+    // Alligns the frame along page size
+    frame_size = _member_alignment(data_size + metadata_size + valid_size, alignment);
+
+    // memory_size should be _ntime * nfreq * file_frame_size (data + metadata)
+    buf_addr = assign_memory(_fname_buf, (structured_data_size * structured_data_num) + (_ntime * nfreq * access_record_size) + (_ntime * nfreq * frame_size));
+
+    // The elements contained in the structured data and access record are each 64 bytes
+    structured_data_addr = (uint64_t*) buf_addr;
+    access_record_addr = (int64_t*) (structured_data_addr + structured_data_num);
+    buf_addr += (structured_data_size * structured_data_num) + (_ntime * nfreq * access_record_size);
+
+    // Record structure of data
```

Comment:
Structs are a language feature for a reason. Define one for the structural data, and then just copy the whole struct. Not only will your code be cleaner, but defining the struct in the header gives a clear statement about what the interface is.